error[E0152]: duplicate lang item in crate `ink_env` for the S3 crate when i try to compile it.

Hi, PhalaTeam !

I'm trying to use the S3 crate with ink4 but i meet perpetually this error `error[E0152]: duplicate lang item in crate "ink_env\" (which `ink` depends on): `panic_impl`.` : ` below is my most simplified version.

 ```
Compiling pink-s3 v0.4.0
error[E0152]: duplicate lang item in crate "ink_env" (which "ink" depends on): "panic_impl".
  |
  = note: the lang item is first defined in crate "std" (which "pink_s3" depends on)
  = note: first definition in "std" loaded from /root/contracts/target/ink/wasm32-unknown-unknown/release/deps/libstd-ed63f69e89d218fa.rmeta
  = note: second definition in "ink_env" loaded from /root/contracts/target/ink/wasm32-unknown-unknown/release/deps/libink_env-3f65290641ea9f9f.rmeta

For more information about this error, try "rustc --explain E0152".
error: could not compile "pink-s3" due to previous error
warning: build failed, waiting for other jobs to finish...
```

```
rustc -V
rustc 1.67.0 (fc594f156 2023-01-24)
cargo contract -V
cargo-contract-contract 2.0.1-unknown-x86_64-unknown-linux-gnu
```

Cargo.toml : 

```
[package]
name = "phat2meet"
version = "0.1.0"
authors = ["PhalaTeamFR-https://linktr.ee/phalafr"]
edition = "2021"

[dependencies]
pink-s3 = { version = "0.4.0", default-features = false }

[dev-dependencies]

[lib]
path = "lib.rs"

[features]
default = ["std"]
std = [
    "pink-s3/std",
]
ink-as-dependency = []
```


lib.rs

```
#![cfg_attr(not(feature = "std"), no_std)]

use pink_s3 as s3;

let endpoint = "s3.ap-southeast-1.amazonaws.com";
let region = "ap-southeast-1";
let access_key = "<Put your S3 access key here>";
let secret_key = "<Put your S3 access secret key here>";

let s3 = s3::S3::new(endpoint, region, access_key, secret_key)
    .unwrap()
    .virtual_host_mode(); // virtual host mode is required for newly created AWS S3 buckets.

let bucket = "my-wallet";
let object_key = "path/to/foo";
let value = b"bar";

s3.put(bucket, object_key, value).unwrap();

let head = s3.head(bucket, object_key).unwrap();
assert_eq!(head.content_length, value.len() as u64);

let v = s3.get(bucket, object_key).unwrap();
assert_eq!(v, value);

s3.delete(bucket, object_key).unwrap();
```

--- a/crates/pink-libs/s3/Cargo.toml
+++ b/crates/pink-libs/s3/Cargo.toml
@@ -3,7 +3,7 @@ description = "Simple S3 client for Phala's pink"
 homepage = "https://github.com/Phala-Network/phala-blockchain"
 license = "Apache-2.0"
 name = "pink-s3"
-version = "0.4.0"
+version = "0.4.1"
 edition = "2021"
 keywords = ["fat-contract", "pink", "ink", "S3"]
 

--- a/crates/pink-libs/s3/example/.gitignore
+++ b/crates/pink-libs/s3/example/.gitignore
@@ -1,0 +1,9 @@
+# Ignore build artifacts from the local tests sub-crate.
+/target/
+
+# Ignore backup files creates by cargo fmt.
+**/*.rs.bk
+
+# Remove Cargo.lock when creating an executable, leave it for libraries
+# More information here http://doc.crates.io/guide.html#cargotoml-vs-cargolock
+Cargo.lock

--- a/crates/pink-libs/s3/example/Cargo.toml
+++ b/crates/pink-libs/s3/example/Cargo.toml
@@ -1,0 +1,28 @@
+[workspace]
+[package]
+name = "example"
+version = "0.1.0"
+authors = ["<author>"]
+edition = "2021"
+
+[dependencies]
+ink = { version = "4", default-features = false }
+pink-s3 = { version = "0.4.0", default-features = false, path = ".." }
+scale = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
+scale-info = { version = "2.3", default-features = false, features = ["derive"], optional = true }
+
+
+[dev-dependencies]
+pink-extension-runtime = { path = "../../../pink/pink-extension-runtime" }
+
+[lib]
+path = "lib.rs"
+
+[features]
+default = ["std"]
+std = [
+    "pink-s3/std",
+    "ink/std",
+    "scale-info/std",
+]
+ink-as-dependency = []

--- a/crates/pink-libs/s3/example/lib.rs
+++ b/crates/pink-libs/s3/example/lib.rs
@@ -1,0 +1,61 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[ink::contract]
+mod example {
+
+    #[ink(storage)]
+    pub struct Example {}
+
+    impl Example {
+        #[ink(constructor)]
+        pub fn default() -> Self {
+            Self {}
+        }
+
+        #[ink(message)]
+        pub fn it_works(&self) {
+            use pink_s3 as s3;
+
+            // I don't care to expose them.
+            let endpoint = "s3.kvin.wang:8443";
+            let region = "garage";
+            let access_key = env!("S3_ACCESS_KEY");
+            let secret_key = env!("S3_SECRET_KEY");
+
+            let s3 = s3::S3::new(endpoint, region, access_key, secret_key)
+                .unwrap()
+                .virtual_host_mode(); // virtual host mode is required for newly created AWS S3 buckets.
+
+            let bucket = "fat-1";
+            let object_key = "path/to/foo";
+            let value = b"bar";
+
+            s3.put(bucket, object_key, value).unwrap();
+
+            let head = s3.head(bucket, object_key).unwrap();
+            assert_eq!(head.content_length, value.len() as u64);
+
+            let v = s3.get(bucket, object_key).unwrap();
+            assert_eq!(v, value);
+
+            s3.delete(bucket, object_key).unwrap();
+        }
+    }
+
+    /// Unit tests in Rust are normally defined within such a `#[cfg(test)]`
+    /// module and test functions are marked with a `#[test]` attribute.
+    /// The below code is technically just normal Rust code.
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+
+        /// We test a simple use case of our contract.
+        #[ink::test]
+        fn it_works() {
+            pink_extension_runtime::mock_ext::mock_all_ext();
+
+            let example = Example::default();
+            example.it_works();
+        }
+    }
+}

--- a/crates/pink-libs/s3/src/lib.rs
+++ b/crates/pink-libs/s3/src/lib.rs
@@ -1,4 +1,8 @@
+#![cfg_attr(not(feature = "std"), no_std)]
 #![doc = include_str!("../README.md")]
+
+#[macro_use]
+extern crate alloc;
 
 use pink::chain_extension::HttpResponse;
 use pink_extension as pink;
@@ -7,6 +11,11 @@ use scale::{Decode, Encode};
 // To encrypt/decrypt HTTP payloads
 
 // To generate AWS4 Signature
+use alloc::{
+    borrow::ToOwned,
+    string::{String, ToString},
+    vec::Vec,
+};
 use hmac::{Hmac, Mac};
 use sha2::Digest;
 use sha2::Sha256;
@@ -123,9 +132,8 @@ impl<'a> S3<'a> {
             format!("/{bucket_name}/{object_key}")
         };
         let canonical_querystring = "";
-        let canonical_headers = format!(
-            "host:{host}\nx-amz-content-sha256:{payload_hash}\nx-amz-date:{amz_date}\n"
-        );
+        let canonical_headers =
+            format!("host:{host}\nx-amz-content-sha256:{payload_hash}\nx-amz-date:{amz_date}\n");
         let signed_headers = "host;x-amz-content-sha256;x-amz-date";
         let canonical_request = format!(
             "{method}\n{canonical_uri}\n{canonical_querystring}\n{canonical_headers}\n{signed_headers}\n{payload_hash}"
@@ -135,9 +143,8 @@ impl<'a> S3<'a> {
         let algorithm = "AWS4-HMAC-SHA256";
         let credential_scope = format!("{datestamp}/{}/{service}/aws4_request", self.region);
         let canonical_request_hash = format!("{:x}", Sha256::digest(canonical_request.as_bytes()));
-        let string_to_sign = format!(
-            "{algorithm}\n{amz_date}\n{credential_scope}\n{canonical_request_hash}"
-        );
+        let string_to_sign =
+            format!("{algorithm}\n{amz_date}\n{credential_scope}\n{canonical_request_hash}");
 
         // 3. Calculate signature
         let signature_key = get_signature_key(
@@ -188,8 +195,10 @@ fn times() -> (String, String) {
     #[cfg(not(test))]
     let datetime = {
         use chrono::{TimeZone, Utc};
-        let time = pink::env().block_timestamp() / 1000;
-        Utc.timestamp(time.try_into().unwrap(), 0)
+        let time = pink::ext().untrusted_millis_since_unix_epoch() / 1000;
+        Utc.timestamp_opt(time as _, 0)
+            .earliest()
+            .expect("Could not convert timestamp to Utc")
     };
 
     // Format both date and datetime for AWS4 signature
@@ -222,39 +231,4 @@ fn get_signature_key(
     let k_region = hmac_sign(&k_date, region_name);
     let k_service = hmac_sign(&k_region, service_name);
     hmac_sign(&k_service, b"aws4_request")
-}
-
-#[cfg(test)]
-mod tests {
-    #[test]
-    #[ignore = "can not run concurrently"]
-    fn it_works() {
-        use crate as s3;
-
-        pink_extension_runtime::mock_ext::mock_all_ext();
-
-        // I don't care to expose them.
-        let endpoint = "s3.kvin.wang:8443";
-        let region = "garage";
-        let access_key = "GKb36294dbfd49a894b19c20cb";
-        let secret_key = "c36c43f1ae5bcb27733753a633fb5df82cc57832822275a761d711637bb268d5";
-
-        let s3 = s3::S3::new(endpoint, region, access_key, secret_key)
-            .unwrap()
-            .virtual_host_mode(); // virtual host mode is required for newly created AWS S3 buckets.
-
-        let bucket = "fat-1";
-        let object_key = "path/to/foo";
-        let value = b"bar";
-
-        s3.put(bucket, object_key, value).unwrap();
-
-        let head = s3.head(bucket, object_key).unwrap();
-        assert_eq!(head.content_length, value.len() as u64);
-
-        let v = s3.get(bucket, object_key).unwrap();
-        assert_eq!(v, value);
-
-        s3.delete(bucket, object_key).unwrap();
-    }
 }

